src/vset.c

```diff
@@ -0,0 +1,2094 @@
+        pv->alloc = required;
+    }
+    return pv;
+}
+
+/**
+ * pvSplit - Splits a pVector into two parts at a given index.
+ *
+ * Arguments:
+ * pv_ptr:       A pointer to the pVector* to split. This pointer is
+ *                updated in-place to point to the left portion (elements [0..split_index-1]).
+ * split_index:  The index at which to split the vector. The resulting right
+ *                vector will contain elements [split_index..len-1].
+ *
+ * This function is used to **efficiently split a sorted vector of pointers**
+ * into two separate vectors. The original vector is truncated in-place to
+ * only contain the first half, and a new vector is returned containing the
+ * second half. This allows for logical partitioning of data without scanning
+ * or reallocating unnecessary memory.
+ *
+ * The vector is assumed to be densely packed and its elements are of type `void*`.
+ *
+ * Memory is allocated for the new right vector using `zmalloc`, and the unused
+ * portion of the original vector may be freed or shrunk via `pvShrinkToFit`
+ * to optimize memory usage.
+ *
+ * Return:
+ *   - A new pVector containing the right split [split_index..len-1].
+ *   - `NULL` in the following cases:
+ *       • The input vector is `NULL`.
+ *       • The input vector has only 1 or fewer elements (nothing to split).
+ *       • The `split_index` is equal to the vector length (all elements stay in the left part).
+ *       • The `split_index` is such that the right part would have 0 elements.
+ *
+ * Side effects:
+ *   - The original vector pointer (`*pv_ptr`) is modified to point to the
+ *     resized left portion.
+ *
+ * Example:
+ * --------
+ * Suppose `pv_ptr` points to a vector of 5 elements:
+ *     [A, B, C, D, E]
+ *
+ * Calling:
+ *     pVector *right = pvSplit(&pv_ptr, 3);
+ *
+ * Results in:
+ *     pv_ptr -> [A, B, C]
+ *     right   -> [D, E]
+ *
+ * If the split_index is 5 (i.e. the end), the function returns NULL and the
+ * original vector is unchanged. */
+pVector *pvSplit(pVector **pv_ptr, uint32_t split_index) {
+    pVector *pv = *pv_ptr;
+
+    // Handle edge cases: null or empty
+    if (!pv || pv->len <= 1) return NULL;
+
+    // If no valid split found, return NULL (entire vector is one block)
+    if (split_index == pv->len) return NULL;
+
+    // Number of elements for the right half
+    uint64_t right_len = pv->len - split_index;
+    if (right_len == 0) return NULL;
+
+    // Allocate new vector for right part
+    size_t item_bytes = sizeof(void *);
+    size_t total_bytes = sizeof(pVector) + right_len * item_bytes;
+    size_t new_alloc;
+    pVector *right = zmalloc_usable(total_bytes, &new_alloc);
+    right->alloc = new_alloc;
+    right->len = right_len;
+
+    // Copy the right part
+    memcpy(&right->data[0], &pv->data[split_index], right_len * item_bytes);
+
+    // Shrink original vector
+    pv->len = split_index;
+    *pv_ptr = pvShrinkToFit(pv); // Optional: shrink in-place to reduce memory
```

Comment:
Comment is wrong/misleading.  If a shrink is performed, it's not "in-place".  Unless we're dealing with very large allocations, the shrink will move the block (or do nothing).
```suggestion
    *pv_ptr = pvShrinkToFit(pv);
```